src/cargo/core/resolver/encode.rs

```diff
@@ -68,36 +163,109 @@ impl EncodableResolve {
                     Some(&source) => PackageId::new(&pkg.name, &pkg.version, source)?,
                 };
 
+                // If a package has a checksum listed directly on it then record
+                // that here, and we also bump our version up to 2 since V1
+                // didn't ever encode this field.
+                if let Some(cksum) = &pkg.checksum {
+                    version = ResolveVersion::V2;
+                    checksums.insert(id, Some(cksum.clone()));
+                }
+
                 assert!(live_pkgs.insert(enc_id, (id, pkg)).is_none())
             }
             live_pkgs
         };
 
-        let lookup_id = |enc_id: &EncodablePackageId| -> Option<PackageId> {
-            live_pkgs.get(enc_id).map(|&(id, _)| id)
-        };
+        // When decoding a V2 version the edges in `dependencies` aren't
+        // guaranteed to have either version or source information. This `map`
+        // is used to find package ids even if dependencies have missing
+        // information. This map is from name to version to source to actual
+        // package ID. (various levels to drill down step by step)
+        let mut map = HashMap::new();
+        for (id, _) in live_pkgs.values() {
+            map.entry(id.name().as_str())
+                .or_insert(HashMap::new())
+                .entry(id.version().to_string())
+                .or_insert(HashMap::new())
+                .insert(id.source_id(), *id);
+        }
 
-        let g = {
-            let mut g = Graph::new();
+        let mut lookup_id = |enc_id: &EncodablePackageId| -> Option<PackageId> {
+            // The name of this package should always be in the larger list of
+            // all packages.
+            let by_version = map.get(enc_id.name.as_str())?;
+
+            // If the version is provided, look that up. Otherwise if the
+            // version isn't provided this is a V2 manifest and we should only
+            // have one version for this name. If we have more than one version
+            // for the name then it's ambiguous which one we'd use. That
+            // shouldn't ever actually happen but in theory bad git merges could
+            // produce invalid lock files, so silently ignore these cases.
```

Comment:
Shouldn't there be (loud) error when merge creates malformed lockfile?

Reply:
Unless you've got a compelling reason to do so, we've got historical rationle for not doing this. In general `Cargo.lock` can be confusing for new users, especially if scary merge messages are happening. What's happening in this case is concurrent updates, and Cargo should handle that the way it does all other updates (aka edits to `Cargo.toml`) which is to automatically, silently, and incrementally update `Cargo.lock`.